File: app/src/main/java/i/am/shiro/amai/adapter/BookAdapter.java
package i.am.shiro.amai.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.annimon.stream.function.Consumer;
import com.annimon.stream.function.IntConsumer;
import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DecodeFormat;

import i.am.shiro.amai.R;
import i.am.shiro.amai.model.Book;
import i.am.shiro.amai.model.Image;

import static android.view.View.INVISIBLE;
import static android.view.View.VISIBLE;
import static java.lang.String.valueOf;

public class BookAdapter extends ListAdapter<Book, BookAdapter.ViewHolder> {

    private final Fragment parentFragment;

    private final LayoutInflater inflater;

    private Consumer<Book> onItemClickListener;

    private Consumer<Book> onItemLongClickListener;

    private IntConsumer onPositionBindListener;

    public BookAdapter(Fragment parentFragment, LayoutInflater inflater) {
        super(new DiffCallback());
        this.parentFragment = parentFragment;
        this.inflater = inflater;
        setHasStableIds(true);
    }

    public void setOnItemClickListener(Consumer<Book> listener) {
        onItemClickListener = listener;
    }

    public void setOnItemLongClickListener(Consumer<Book> listener) {
        onItemLongClickListener = listener;
    }

    public void setOnPositionBindListener(IntConsumer listener) {
        onPositionBindListener = listener;
    }

    @Override
    public long getItemId(int position) {
        return getItem(position).getId();
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.item_staggered_book, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Book book = getItem(position);
        holder.bind(book);
        if (onPositionBindListener != null) onPositionBindListener.accept(position);
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private final ConstraintLayout constraintLayout;

        private final ImageView thumbnailImage;

        private final TextView titleText;

        private final TextView pageText;

        private final ImageView savedBadge;

        private Book book;

        ViewHolder(View view) {
            super(view);
            constraintLayout = view.findViewById(R.id.constraintLayout);
            thumbnailImage = view.findViewById(R.id.thumbnailImage);
            titleText = view.findViewById(R.id.titleText);
            pageText = view.findViewById(R.id.pageText);
            savedBadge = view.findViewById(R.id.badgeSaved);
            view.setOnClickListener(v -> onItemClick());
            view.setOnLongClickListener(v -> onItemLongClick());
        }

        private void onItemClick() {
            onItemClickListener.accept(book);
        }

        private boolean onItemLongClick() {
            if (onItemLongClickListener != null) {
                onItemLongClickListener.accept(book);
                return true;
            }
            return false;
        }

        private void bind(Book book) {
            this.book = book;

            savedBadge.setVisibility(book.isDownloaded() ? VISIBLE : INVISIBLE);
            titleText.setText(book.getTitle());
            pageText.setText(valueOf(book.getPageCount()));

            Image coverThumbnailImage = book.getCoverThumbnailImage();

            String ratioStr = String.format("h,%s:%s",
                    coverThumbnailImage.getWidth(),
                    coverThumbnailImage.getHeight());

            ConstraintSet constraintSet = new ConstraintSet();
            constraintSet.clone(constraintLayout);
            constraintSet.setDimensionRatio(R.id.thumbnailImage, ratioStr);
            constraintSet.applyTo(constraintLayout);

            Glide.with(parentFragment)
                .load(coverThumbnailImage.getUrl())
                .format(DecodeFormat.PREFER_RGB_565)
                .into(thumbnailImage);
        }
    }

    private static class DiffCallback extends DiffUtil.ItemCallback<Book> {

        @Override
        public boolean areItemsTheSame(@NonNull Book oldItem, @NonNull Book newItem) {
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull Book oldItem, @NonNull Book newItem) {
            return oldItem.isDownloaded() == newItem.isDownloaded();
        }
    }
}

File: app/src/main/java/i/am/shiro/amai/adapter/BookAdapter.kt
package i.am.shiro.amai.adapter

import android.view.View
import android.view.View.INVISIBLE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.view.updateLayoutParams
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DecodeFormat
import i.am.shiro.amai.R
import i.am.shiro.amai.model.Book
import i.am.shiro.amai.util.inflateChild

class BookAdapter(
        private val parentFragment: Fragment,
        private val onItemClick: (Book) -> Unit,
        private val onItemLongClick: (Book) -> Unit = {},
        private val onPositionBind: (Int) -> Unit = {}
) : ListAdapter<Book, BookAdapter.ViewHolder>(DiffCallback()) {

    init {
        setHasStableIds(true)
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).id.toLong()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(parent.inflateChild(R.layout.item_staggered_book))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(position)
        onPositionBind(position)
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {

        private val thumbnailImage = view.findViewById<ImageView>(R.id.thumbnailImage)

        private val titleText = view.findViewById<TextView>(R.id.titleText)

        private val pageText = view.findViewById<TextView>(R.id.pageText)

        private val savedBadge = view.findViewById<ImageView>(R.id.badgeSaved)

        private lateinit var book: Book

        init {
            view.setOnClickListener { onItemClick(book) }
            view.setOnLongClickListener { onItemLongClick() }
        }

        private fun onItemLongClick(): Boolean {
            onItemLongClick(book)
            return true
        }

        fun bind(position: Int) {
            book = getItem(position)

            savedBadge.visibility = if (book.isDownloaded) VISIBLE else INVISIBLE
            titleText.text = book.title
            pageText.text = book.pageCount.toString()

            val coverThumbnailImage = book.coverThumbnailImage

            thumbnailImage.updateLayoutParams<ConstraintLayout.LayoutParams> {
                val width = coverThumbnailImage.width
                val height = coverThumbnailImage.height
                dimensionRatio = "h,$width:$height"
            }

            Glide.with(parentFragment)
                    .load(coverThumbnailImage.url)
                    .format(DecodeFormat.PREFER_RGB_565)
                    .into(thumbnailImage)
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<Book>() {

        override fun areItemsTheSame(oldItem: Book, newItem: Book): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Book, newItem: Book): Boolean {
            return oldItem.isDownloaded == newItem.isDownloaded
        }
    }
}
